appengine/flexible/TwelveFactor/Services/CloudStorageFileProvider.cs

```diff
@@ -0,0 +1,259 @@
+            };
+        }
+
+        /// <summary> Gets the Object at the path.  If it's not found,
+        /// doesn't throw an exception; instead returns null. </summary>
+        Google.Apis.Storage.v1.Data.Object GetObject(ObjectPath path) {
+            try 
+            {
+                return _storage.GetObject(path.BucketName, 
+                    path.ObjectName);
+            }
+            catch (Google.GoogleApiException e)
+            when (e.HttpStatusCode == HttpStatusCode.NotFound) 
+            {
+                _logger.LogWarning("{0}/{1} not found.", path.BucketName,
+                    path.ObjectName);
+                return null;
+            }
+        }
+
+        /// <summary> Watch a file for changes.
+        public IChangeToken Watch(string filter)
+        {
+            if (_pollingInterval == null) {
+                return null;
+            }
+            _logger.LogDebug("Watching {0}", filter);
+            var path = SplitObjectPath(filter);
+            var token = new CloudStorageChangeToken();
+            var obj = GetObject(path);
+            long? generation = obj?.Generation;
+            Task.Run(async () => {
+                while (true) {
+                    await Task.Delay(_pollingInterval.Value);
+                    try
+                    {
+                        obj = GetObject(path);
+                        if (generation != obj?.Generation) {
+                            token.HasChanged = true;
+                            return;
+                        }
+                    } 
+                    catch (Exception e) 
+                    {
+                        _logger.LogError(0, e, "Exception while watching {0}",
+                            filter);
+                    }
+                }
+            });
+            return token;            
+        }            
+    }
+
+    class CloudStorageChangeToken : IChangeToken
+    {
+        List<ChangeCallback> _callbacks = new List<ChangeCallback>();
+        bool _hasChanged = false;
+
+        public bool HasChanged { 
+            get { return _hasChanged; }
+            set {
+                _hasChanged = true;
+                foreach (var callback in _callbacks) {
+                    callback.Invoke();
+                }
+            }
+        }
+
+        public bool ActiveChangeCallbacks => true;
+
+        public IDisposable RegisterChangeCallback(Action<object> callback,
+            object state)
+        {
+            var disposable = new ChangeCallback() {
+                Callback = callback,
+                State = state
+            };
+            _callbacks.Add(disposable);
+            return disposable;
+        }
+
+        class ChangeCallback : IDisposable {
+            public Action<object> Callback { get; set; }
+            public object State {get; set;}
+
+            private object _lock = new object();
+            private bool _done = false;

```

Comment:
Nit, instead of having a boolean `_done` you could just null out the `Callback` property when you raise it. That will make your implementation of `Dispose` below make more sense.